// homebridge-http-base/dist/notifications/pulltimer.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils = __importStar(require("../utils"));
class PullTimer {
    constructor(log, interval, handler, successHandler) {
        this.log = log;
        this.interval = interval;
        this.handler = handler;
        this.successHandler = successHandler;
    }
    start() {
        if (!this.timeout) {
            this.timeout = setTimeout(this.handleTimer.bind(this), this.interval);
        }
        else {
            this.timeout.refresh();
        }
    }
    resetTimer() {
        if (!this.timeout)
            return;
        this.timeout.refresh();
    }
    stop() {
        if (this.timeout) {
            clearTimeout(this.timeout);
        }
        this.timeout = undefined;
    }
    handleTimer() {
        this.handler(utils.once((error, value) => {
            if (error)
                this.log("Error occurred while pulling update from switch: " + error.message);
            else
                this.successHandler(value);
            this.resetTimer();
        }));
    }
}
exports.PullTimer = PullTimer;
